File: lesson_12/structs_java/structs_app/src/main/java/com/codedifferently/lesson12/Stack.java
package com.codedifferently.lesson12;

public class Stack {
  private ListNode top;

  public Stack() {
    this.top = null;
  }

  public void push(int value) {
    ListNode newNode = new ListNode(value);
    newNode.next = top;
    top = newNode;
  }

  public int pop() {
    if (isEmpty()) {
      throw new RuntimeException("Stack Underflow: Cannot pop from an empty stack");
    }
    int poppedValue = top.val;
    top = top.next;
    return poppedValue;
  }

  public int peek() {
    if (isEmpty()) {
      throw new RuntimeException("Stack is empty: Cannot peek");
    }
    return top.val;
  }

  public boolean isEmpty() {
    return top == null;
  }

  public void printStack() {
    ListNode current = top;
    while (current != null) {
      System.out.print(current.val + " -> ");
      current = current.next;
    }
    System.out.println("null");
  }
}

File: lesson_12/structs_ts/src/stack.ts
import { ListNode } from './list_node.js';

export class Stack {
  private top: ListNode | null;

  constructor() {
    this.top = null;
  }

  push(value: number): void {
    const newNode = new ListNode(value, this.top ?? undefined);
    this.top = newNode;
  }

  pop(): number {
    if (this.isEmpty()) {
      throw new Error('Stack Underflow: Cannot pop from an empty stack');
    }
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const poppedValue = this.top!.val;
    this.top = this.top?.next ?? null;
    return poppedValue;
  }

  peek(): number {
    if (this.isEmpty()) {
      throw new Error('Stack is empty: Cannot peek');
    }
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return this.top!.val;
  }

  isEmpty(): boolean {
    return this.top === null;
  }
}
